app/workers/attachments/finish_direct_upload_job.rb

```diff
@@ -29,115 +29,44 @@
 class Attachments::FinishDirectUploadJob < ApplicationJob
   queue_with_priority :high
 
-  def perform(attachment_id, whitelist: true)
+  def perform(attachment_id, allowlist: true)
     attachment = Attachment.pending_direct_upload.find_by(id: attachment_id)
-    # An attachment is guaranteed to have a file.
-    # But if the attachment is nil the expression attachment&.file will be nil and attachment&.file.local_file
-    # will throw a NoMethodError: undefined method local_file' for nil:NilClass`.
-    local_file = attachment && attachment.file.local_file
 
-    if local_file.nil?
-      return Rails.logger.error("File for attachment #{attachment_id} was not uploaded.")
-    end
-
-    User.execute_as(attachment.author) do
-      attach_uploaded_file(attachment, local_file, whitelist)
+    Attachments::FinishDirectUploadService
+      .new(user: attachment.author, model: attachment, contract_options: derive_contract_options(allowlist))
+      .call
+      .on_failure do |call|
+      destroy_attachment_and_log_errors(attachment, call.errors)
     end
   end
 
   private
 
-  def attach_uploaded_file(attachment, local_file, whitelist)
-    set_attributes_from_file(attachment, local_file)
-    validate_attachment(attachment, whitelist)
-    save_attachment(attachment)
-    journalize_container(attachment)
-    attachment_created_event(attachment)
-    schedule_jobs(attachment)
-  rescue StandardError => e
-    ::OpenProject.logger.error e
+  def destroy_attachment_and_log_errors(attachment, errors)
     attachment.destroy
-  ensure
-    FileUtils.rm_rf(local_file.path)
-  end
-
-  def set_attributes_from_file(attachment, local_file)
-    attachment.extend(OpenProject::ChangedBySystem)
-    attachment.change_by_system do
-      attachment.status = :uploaded
-      attachment.set_file_size local_file
-      attachment.set_content_type local_file
-      attachment.set_digest local_file
-    end
-  end
-
-  def save_attachment(attachment)
-    attachment.save! if attachment.changed?
-  end
-
-  def validate_attachment(attachment, whitelist)
-    contract = create_contract attachment, whitelist
-
-    unless contract.valid?
-      errors = contract.errors.full_messages.join(", ")
-      raise "Failed to validate attachment #{attachment.id}: #{errors}"
-    end
-  end
-
-  def create_contract(attachment, whitelist)
-    options = derive_contract_options(whitelist)
-    ::Attachments::CreateContract.new attachment,
-                                      attachment.author,
-                                      options:
-  end
-
-  def schedule_jobs(attachment)
-    attachment.extract_fulltext
+    errors = errors.full_messages
+
+    OpenProject.logger.error(
+      <<~MSG
+        Failed to finish attachment upload for:
+          * user: #{attachment.author}
+          * container: #{attachment.container}
+          * attachment file name: #{attachment.filename}
+
+        Errors:
+          #{errors.join("\n")}
```

Comment:
I suggest indenting all errors, as otherwise all the niceness goes in vain
```suggestion
          #{errors.join("\n          ")}
```

Reply:
Amended, thanks.